source/extensions/transport_sockets/alts/tsi_frame_protector.cc

```diff
@@ -0,0 +1,78 @@
+#include "extensions/transport_sockets/alts/tsi_frame_protector.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+// TODO(lizan): tune size later
+static constexpr uint32_t BUFFER_SIZE = 16384;
+
+TsiFrameProtector::TsiFrameProtector(CFrameProtectorPtr&& frame_protector)
+    : frame_protector_(std::move(frame_protector)) {}
+
+tsi_result TsiFrameProtector::protect(Buffer::Instance& input, Buffer::Instance& output) {
+  ASSERT(frame_protector_);
+
+  unsigned char protected_buffer[BUFFER_SIZE];
+  while (input.length() > 0) {
+    auto* message_bytes = reinterpret_cast<unsigned char*>(input.linearize(input.length()));
+    size_t protected_buffer_size = BUFFER_SIZE;
+    size_t processed_message_size = input.length();
+    tsi_result result =
+        tsi_frame_protector_protect(frame_protector_.get(), message_bytes, &processed_message_size,
+                                    protected_buffer, &protected_buffer_size);
+    if (result != TSI_OK) {
+      ASSERT(result != TSI_INVALID_ARGUMENT && result != TSI_UNIMPLEMENTED);
+      return result;
+    }
+    output.add(protected_buffer, protected_buffer_size);
+    input.drain(processed_message_size);
+  }
+
+  // TSI may buffer some of the input in its internal, flushing its buffer to protected_buffer to
+  // avoid some of input is not written to socket.
+  size_t still_pending_size;
```

Comment:
// TSI may buffer some of the input internally.  Flush its buffer to protected_buffer.
?